base/methodshow.jl

```diff
@@ -11,9 +11,16 @@ function strip_gensym(sym)
 end
 
 function argtype_decl(env, n, @nospecialize(sig::DataType), i::Int, nargs, isva::Bool) # -> (argname, argtype)
-    t = sig.parameters[i]
-    if i == nargs && isva && !isvarargtype(t)
-        t = Vararg{t,length(sig.parameters)-nargs+1}
+    t = fieldtype(sig, i)
```

Comment:
This may give the type of the fields, not the type of the parameters. The former might be normalized in various ways that differ from the argument types specified in the latter (For example, the isva handling below was one such example. Another is that TypeVars may be removed). I agree more general handling is needed though, particularly surrounding the leading/trailing Vararg{N} handling.

Reply:
so you're proposing `t = sig.parameters[unwrapva(min(i, end))]` here essentially?

Reply:
Could you elaborate on the normalization you're thinking of. I was thinking for tuple, they should be basically equivalent.

Reply:
I guess the bigger problem is that sig has free typevars, so passing it to fieldtype is a bit of an abuse. We do that in a few places already, but we should probably avoid introducing more. I'll switch it back.